volumes: data race in Test_Run_OneVolumeAttachAndDetachMultipleNodesWithReadWriteOnce

/kind bug

Seeing this data race when running `go test`:

```
=== RUN   Test_Run_OneVolumeAttachAndDetachMultipleNodesWithReadWriteOnce
==================
WARNING: DATA RACE
Read at 0x00c420284780 by goroutine 43:
  runtime.mapiterinit()
      /usr/lib/golang/src/runtime/hashmap.go:668 +0x0
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/cache.(*desiredStateOfWorld).GetVolumesToAttach()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/cache/desired_state_of_world.go:372 +0x26d
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.Test_Run_OneVolumeAttachAndDetachMultipleNodesWithReadWriteOnce()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/reconciler_test.go:469 +0x124c
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:657 +0x107

Previous write at 0x00c420284780 by goroutine 44:
  runtime.mapassign()
      /usr/lib/golang/src/runtime/hashmap.go:485 +0x0
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/cache.(*desiredStateOfWorld).SetMultiAttachError()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/cache/desired_state_of_world.go:346 +0x2a1
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.(*reconciler).attachDesiredVolumes()
      github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/_test/_obj_test/reconciler.go:274 +0xba8
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.(*reconciler).reconcile()
      github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/_test/_obj_test/reconciler.go:231 +0x23b
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.(*reconciler).reconciliationLoopFunc.func1()
      github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/_test/_obj_test/reconciler.go:73 +0x78
  github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97 +0x6f
  github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:98 +0xcd
  github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait.Until()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52 +0x5a
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.(*reconciler).Run()
      github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/_test/_obj_test/reconciler.go:65 +0x94

Goroutine 43 (running) created at:
  testing.(*T).Run()
      /usr/lib/golang/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /usr/lib/golang/src/testing/testing.go:882 +0xaa
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:657 +0x107
  testing.runTests()
      /usr/lib/golang/src/testing/testing.go:888 +0x4e0
  testing.(*M).Run()
      /usr/lib/golang/src/testing/testing.go:822 +0x1c3
  main.main()
      github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/_test/_testmain.go:100 +0x33d

Goroutine 44 (running) created at:
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.Test_Run_OneVolumeAttachAndDetachMultipleNodesWithReadWriteOnce()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/reconciler_test.go:449 +0x10bf
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:657 +0x107
==================
==================
WARNING: DATA RACE
Read at 0x00c420296268 by goroutine 43:
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/cache.(*desiredStateOfWorld).GetVolumesToAttach()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/cache/desired_state_of_world.go:372 +0x2bc
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.Test_Run_OneVolumeAttachAndDetachMultipleNodesWithReadWriteOnce()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/reconciler_test.go:469 +0x124c
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:657 +0x107

Previous write at 0x00c420296268 by goroutine 44:
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/cache.(*desiredStateOfWorld).SetMultiAttachError()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/cache/desired_state_of_world.go:346 +0x2c0
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.(*reconciler).attachDesiredVolumes()
      github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/_test/_obj_test/reconciler.go:274 +0xba8
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.(*reconciler).reconcile()
      github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/_test/_obj_test/reconciler.go:231 +0x23b
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.(*reconciler).reconciliationLoopFunc.func1()
      github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/_test/_obj_test/reconciler.go:73 +0x78
  github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97 +0x6f
  github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:98 +0xcd
  github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait.Until()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52 +0x5a
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.(*reconciler).Run()
      github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/_test/_obj_test/reconciler.go:65 +0x94

Goroutine 43 (running) created at:
  testing.(*T).Run()
      /usr/lib/golang/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /usr/lib/golang/src/testing/testing.go:882 +0xaa
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:657 +0x107
  testing.runTests()
      /usr/lib/golang/src/testing/testing.go:888 +0x4e0
  testing.(*M).Run()
      /usr/lib/golang/src/testing/testing.go:822 +0x1c3
  main.main()
      github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/_test/_testmain.go:100 +0x33d

Goroutine 44 (running) created at:
  github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler.Test_Run_OneVolumeAttachAndDetachMultipleNodesWithReadWriteOnce()
      /go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler/reconciler_test.go:449 +0x10bf
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:657 +0x107
==================
--- FAIL: Test_Run_OneVolumeAttachAndDetachMultipleNodesWithReadWriteOnce (0.08s)
	reconciler_test.go:520: Warning: Wrong NewAttacherCallCount. Expected: <1> Actual: <0>. Will retry.
	reconciler_test.go:547: Warning: Wrong NewDetacherCallCount. Expected: <1> Actual: <0>. Will retry.
	reconciler_test.go:547: Warning: Wrong NewDetacherCallCount. Expected: <1> Actual: <0>. Will retry.
	reconciler_test.go:547: Warning: Wrong NewDetacherCallCount. Expected: <1> Actual: <0>. Will retry.
	testing.go:610: race detected during execution of test
FAIL
```

--- a/pkg/controller/volume/attachdetach/cache/desired_state_of_world.go
+++ b/pkg/controller/volume/attachdetach/cache/desired_state_of_world.go
@@ -336,8 +336,8 @@ func (dsw *desiredStateOfWorld) VolumeExists(
 func (dsw *desiredStateOfWorld) SetMultiAttachError(
 	volumeName v1.UniqueVolumeName,
 	nodeName k8stypes.NodeName) {
-	dsw.RLock()
-	defer dsw.RUnlock()
+	dsw.Lock()
+	defer dsw.Unlock()
 
 	nodeObj, nodeExists := dsw.nodesManaged[nodeName]
 	if nodeExists {

